util/template/expression_template.go

```diff
@@ -21,53 +21,25 @@ func init() {
 }
 
 func expressionReplace(w io.Writer, expression string, env map[string]interface{}, allowUnresolved bool) (int, error) {
-	// The template is JSON-marshaled. This JSON-unmarshals the expression to undo any character escapes.
-	var unmarshalledExpression string
-	err := json.Unmarshal([]byte(fmt.Sprintf(`"%s"`, expression)), &unmarshalledExpression)
-	if err != nil && allowUnresolved {
-		log.WithError(err).Debug("unresolved is allowed ")
-		return w.Write([]byte(fmt.Sprintf("{{%s%s}}", kindExpression, expression)))
-	}
-	if err != nil {
-		return 0, fmt.Errorf("failed to unmarshall JSON expression: %w", err)
-	}
-
-	if _, ok := env["retries"]; !ok && hasRetries(unmarshalledExpression) && allowUnresolved {
-		// this is to make sure expressions like `sprig.int(retries)` don't get resolved to 0 when `retries` don't exist in the env
-		// See https://github.com/argoproj/argo-workflows/issues/5388
-		log.WithError(err).Debug("Retries are present and unresolved is allowed")
-		return w.Write([]byte(fmt.Sprintf("{{%s%s}}", kindExpression, expression)))
-	}
-
-	// This is to make sure expressions which contains `workflow.status` and `work.failures` don't get resolved to nil
-	// when `workflow.status` and `workflow.failures` don't exist in the env.
-	// See https://github.com/argoproj/argo-workflows/issues/10393, https://github.com/expr-lang/expr/issues/330
-	// This issue doesn't happen to other template parameters since `workflow.status` and `workflow.failures` only exist in the env
-	// when the exit handlers complete.
-	if ((hasWorkflowStatus(unmarshalledExpression) && !hasVarInEnv(env, "workflow.status")) ||
-		(hasWorkflowFailures(unmarshalledExpression) && !hasVarInEnv(env, "workflow.failures"))) &&
-		allowUnresolved {
-		return w.Write([]byte(fmt.Sprintf("{{%s%s}}", kindExpression, expression)))
+	if isTernaryExpression(expression) {
+		splits := strings.SplitN(expression, "?", 2)
+		cond := splits[0]
+		result, err := resolveExpression(cond, env, allowUnresolved)
```

Comment:
doing a partial eval here sounds like it could potentially cause a whole lot of issues, I don't think we should be splitting up expressions like that as `expr` was not meant to be used that way. Special casing ternaries also feels potentially problematic.

could you solve this a different way without either of those?

Reply:
Yes, I think split is not a good solution, may causing other issues. May be we can report correct error for user to debug, if have "interface conversion" error when resolveExpression.